driver/src/price_estimation/dexag/mod.rs

```diff
@@ -1 +1,178 @@
+                STABLE_COIN
+            )
+        })?;
+
+        Ok(Self {
+            api,
+            tokens,
+            stable_coin,
+        })
+    }
+}
+
+impl<Api> PriceSource for DexagClient<Api>
+where
+    Api: DexagApi,
+{
+    fn get_prices(&self, tokens: &[Token]) -> Result<HashMap<TokenId, u128>> {
+        Ok(tokens
+            .iter()
+            .filter_map(|token| {
+                let stable_coin_price = if token.symbol() == self.stable_coin.symbol {
+                    1.0
+                } else {
+                    let dexag_token = self.tokens.get(token.symbol())?;
+                    self.api.get_price(&self.stable_coin, dexag_token).ok()?
+                };
+                Some((token.id, token.get_owl_price(stable_coin_price)))
+            })
+            .collect())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::api::MockDexagApi;
+    use super::*;
+    use mockall::predicate::*;
+
+    #[test]
+    fn get_token_prices() {
+        let mut api = MockDexagApi::new();
+
+        let tokens = [
+            Token::new(1, "ETH", 18),
+            Token::new(4, "USDC", 6),
+            Token::new(5, "PAX", 18),
+            Token::new(6, "DAI", 18),
+        ];
+
+        let api_tokens = [
+            super::api::Token {
+                name: String::new(),
+                symbol: "DAI".to_string(),
+                address: None,
+            },
+            super::api::Token {
+                name: String::new(),
+                symbol: "ETH".to_string(),
+                address: None,
+            },
+            super::api::Token {
+                name: String::new(),
+                symbol: "USDC".to_string(),
+                address: None,
+            },
+            super::api::Token {
+                name: String::new(),
+                symbol: "PAX".to_string(),
+                address: None,
+            },
+        ];
+
+        let api_tokens_ = api_tokens.clone();
+        api.expect_get_token_list()
+            .returning(move || Ok(api_tokens_.to_vec()));
+
+        api.expect_get_price()
+            .with(eq(api_tokens[0].clone()), eq(api_tokens[1].clone()))
+            .returning(|_, _| Ok(0.7));
+        api.expect_get_price()
+            .with(eq(api_tokens[0].clone()), eq(api_tokens[2].clone()))
+            .returning(|_, _| Ok(1.2));
+        api.expect_get_price()
+            .with(eq(api_tokens[0].clone()), eq(api_tokens[3].clone()))
+            .returning(|_, _| Err(anyhow!("")));
+        api.expect_get_price()
+            .with(
+                eq(api_tokens[0].clone()),
+                // Clippy claims we can remove the clone here but compilation
+                // actually fails if we follow the suggestion.
+                #[allow(clippy::redundant_clone)]
+                eq(api_tokens[0].clone()),
+            )
+            .returning(|_, _| Ok(1.0));
+
+        let client = DexagClient::with_api(api).unwrap();
```

Comment:
Should we test what happens with an API that doesn't return DAI?

Reply:
Added a test for it.